File: linked_list/2-linked_list_add_delete.js
// Write code to add, search and remove items from an unsorted linked list.

class LinkedList {

    constructor () {
        this.head = null
        this.size = 0
    }

    dump () {
        console.log(`Head is ${this.head.cargo}, size is ${this.size}`)
    }

    add (cargo) {
        let node = new Node(cargo, this.head)
        this.head = node
        this.size += 1
        return this.head
    }

    search (cargo) {

        if (!this.head) return null

        let node = this.head

        while (node) {
            if (node.cargo === cargo) {
                console.log(`Found ${cargo}`)
                return node
            }
            node = node.next
        }

        return `node with cargo ${v} not found`

    }

    delete (cargo) {

        if (!this.head) return null

        let node = this.head

        if (node.cargo === cargo) {
            this.head = node.next
            this.size -= 1
            console.log(`Deleted ${cargo}`)
            return
        }

        while (node.next) {
            if (node.next.cargo === cargo) {
                node.next = node.next.next
                this.size -= 1
                console.log(`Deleted ${cargo}`)
                return
            }
            node = node.next
        }

    }

    print () {
        console.log(`---------- printing`)
        if (this.head) {
            this.head.print()
        }
    }

}

class Node {

    constructor (cargo, next) {
        this.cargo = cargo
        this.next = next
    }

    print () {
        console.log(this.cargo)
        if (this.next) {
            this.next.print()
        }
    }

}

let l = new LinkedList()

l.add('1')
l.add('2')
l.add('3')
l.add('4')
l.add('5')
l.add('6')

l.dump()

l.print()

l.search('2')

l.delete('2')

l.print()

File: linked_list/2-linked_list_add_delete.py
"""
Write code to add, search and remove items from an unsorted linked list.
"""

class LinkedList:

    def __init__(self):
        self.length = 0
        self.head = None

    def add(self, cargo):
        _next = self.head
        self.head = Node(cargo, _next)
        self.length += 1

    def search(self, value):
        node = self.head
        while node:
            if node.cargo == value:
                return node
            node = node.next
        return None

    def delete(self, k):
        node = self.head

        # The node to delete is the head, change it.
        if node.cargo == k:
            self.head = node.next
            del node
            self.length -= 1
            return

        # The node to delete is not the head, update the chain.
        while node.next:
            if node.next.cargo == k:
                node.next = node.next.next
                del node.next
                self.length -= 1
                return
            node = node.next

    def print_forward(self):
        print('[', end='')
        if self.head:
            self.head.print_forward()
        print(']')


class Node:

    def __init__(self, cargo=None, _next=None):
        self.cargo = cargo
        self.next = _next

    def print_forward(self):
        print(self, end=',' if self.next else '')
        if self.next:
            self.next.print_forward()

    def __str__(self):
        return str(self.cargo)


linked_list = LinkedList()

linked_list.add(1)
linked_list.add(2)
linked_list.add(3)

print('-' * 40)
print(f'Head = {linked_list.head.cargo}')
print(f'Length = {linked_list.length}')
linked_list.print_forward()

print('-' * 40)
linked_list.delete(3)
print(f'Head = {linked_list.head.cargo}')
print(f'Length = {linked_list.length}')
linked_list.print_forward()
